src/security/PeerConnector.cc

```diff
@@ -75,12 +79,24 @@ Security::PeerConnector::commCloseHandler(const CommCloseCbParams &params)
     connectionClosed("Security::PeerConnector::commCloseHandler");
 }
 
+void
+Security::PeerConnector::commTimeoutHandler(const CommTimeoutCbParams &)
+{
+    debugs(83, 5, serverConnection() << " timedout. this=" << (void*)this);
+    auto anErr = new ErrorState(ERR_SECURE_CONNECT_FAIL, Http::scGatewayTimeout, request.getRaw(), al);
+    anErr->detail = new Ssl::ErrorDetail(SQUID_ERR_SSL_HANDSHAKE, nullptr, nullptr);
+    bail(anErr);
+    mustStop("Timedout");
+}
+
 void
 Security::PeerConnector::connectionClosed(const char *reason)
 {
     debugs(83, 5, reason << " socket closed/closing. this=" << (void*)this);
+    auto anErr = new ErrorState(ERR_SECURE_CONNECT_FAIL, Http::scServiceUnavailable, request.getRaw(), al);
+    anErr->detail = new Ssl::ErrorDetail(SQUID_ERR_SSL_HANDSHAKE, nullptr, nullptr);
+    bail(anErr);
     mustStop(reason);
```

Comment:
I have not checked all the cases, but it feels like we are always calling mustStop() after calling bail() or bailWith(). Please check and, if so, 
* either remove these mustStop() calls (if doneAll() already returns true after errors or can be easily adjusted to return true after errors)
* or move them (with some generic argument like "bailed") into the bail*() method (otherwise).

Reply:
The mustStop() is not needed after bail() because bail makes the callback member nil (after schedule the callback) so the doneAll() returns true.
I removed these calls.

commit fb6304d47a for Security::PeerConnector
commit f0dedcd0f7 fr Http::Tunneler